code/contracts/ONEWallet2.sol

```diff
@@ -0,0 +1,176 @@
+                if(operType != OperType.RECOVERY){                                        
+                    _verifyMerkleProof(neighbors, otpIdx, eotp);
+                } else {
+                    require(keccak256(abi.encodePacked(eotp)) == recoveryHash, "Provided secret for recovery is incorrect."); // here is also OK to revert since invalid secret (in eotp) was submitted
+                }
+
+                // 4) execute the operation according to its type; never revert since a valid OTP + proof are already submitted in args.
+                if(operType == OperType.TRANSFER){                    
+
+                    if(!_executeTransfer(addr, amount)){
+                        lastRevealedOTPIdx = otpIdx; // do not allow to call reveal with this eotp again
+                        return false; // if daily limit is not met or there was an error with send()
+                    }                
+                    _cleanUpDailySpendings(); // clean up only after a successful transfer
+
+                } else if(operType == OperType.SET_LAST_RESORT_ADDR){
+                    if(lastResortAddress == address(0x0) && addr != address(0x0)) // alow setting up last resort address only if it was 0x0 before
+                        lastResortAddress = addr;
+                } else if(operType == OperType.RECOVERY){ // we do not care about params in Commits
+                    _drain(); // IH: I do not find any reason for this operation to be executed using OTP.
+                } else {
+                    emit UnknownOperationType(operType);                    
+                }                
+                
+                delete commits[otpIdx];
+                lastRevealedOTPIdx = otpIdx; // it is not possible to execute any operations using the otpIdx anymore (handled in commit)
+                return true;
+            }
+        }
+        return false;        
+    }
+
+    //////////////////////// Public Aux Functions ////////////////////////// 
+
+    /**
+     * It serves for the deletion of accidentaly submitted commits by the user or malicious commits submitted by the attackers, which contribute to the total storage allocation.
+     * Currently, this method can be called by anybody since it does touch only the garbage. 
+     * Also, in theory, someone may earn gas if there are too many entries submitted by the attacker (who payed for it).
+     */
+    function cleanUpCommits() external {                 
+        for (uint32 i = lastCleanUpOfCommitsIdx; i <= lastRevealedOTPIdx; i++){ // IH: in theory, this might lead to out-of-gas exception for settings of TOTP with very long validity but currently I'd not be affraid. Moreover, we are  getting the gas back for successfull deletions.
+            delete commits[i]; // NOTE: I'd say it makes no difference whether it was initialized or not, so we can delete any entry.        
```

Comment:
For primitive arrays, a deletion on array itself is sufficient to delete everything. But I am uncertain whether it would be sufficient for struct arrays.

Reply:
The compiler did not complain about it.

Reply:
The spec reads:
> `delete a` assigns the initial value for the type to `a`. I.e. for integers it is equivalent to `a = 0`, but it can also be used on arrays, where it assigns a dynamic array of length zero or a static array of the same length with all elements set to their initial value
(https://docs.soliditylang.org/en/v0.8.4/types.html#delete)

It is unclear from the language whether a dynamic array's elements would be set to their initial value